### PR TITLE
Make Registration form a live component and enable tests for login, logout and account verification

### PR DESCRIPTION
**Pull request by TheMilek**
| Q               | A
|-----------------|-----
| Branch?         | bootstrap-shop 
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| License         | MIT

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/spec/Form/EventSubscriber/CustomerRegistrationFormSubscriberSpec.php [hunk 1]
```diff
@@ -62,16 +62,18 @@ function it_sets_user_for_existing_customer(
         $this->preSubmit($event);
     }
 
-    function it_throws_unexpected_type_exception_if_data_is_not_customer_type(
+    function it_does_nothing_if_data_is_not_customer_type(
         FormEvent $event,
         FormInterface $form,
         ShopUserInterface $user,
+        RepositoryInterface $customerRepository,
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/spec/Form/EventSubscriber/CustomerRegistrationFormSubscriberSpec.php]: As this is an argument of subscriber's constructor, it should be added as the first one here

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/account/register.yaml [hunk 2]
```diff
@@ -1,24 +1,27 @@
...
-            subtitle:
-                template: '@SyliusShop/account/register/header/subtitle.html.twig'
+                component: 'sylius_shop:register:form'
+                props:
+                    resource: '@=_context.customer'
+                    form: '@=_context.form'
 
         'sylius_shop.account.register.form':
+            header:
+                template: '@SyliusShop/account/register/header.html.twig'
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/account/register.yaml]:
```suggestion
                template: '@SyliusShop/account/register/form/header.html.twig'
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/account/register.yaml [hunk 3]
```diff
@@ -1,24 +1,27 @@
...
             credentials:
                 template: '@SyliusShop/account/register/form/credentials.html.twig'
             buttons:
                 template: '@SyliusShop/account/register/form/buttons.html.twig'
+                
+        'sylius_shop.account.register.form.header':
+            title:
+                template: '@SyliusShop/account/register/header/title.html.twig'
+            subtitle:
+                template: '@SyliusShop/account/register/header/subtitle.html.twig'
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/config/app/twig_hooks/account/register.yaml]:
```suggestion
        'sylius_shop.account.register.form.header':
            title:
                template: '@SyliusShop/account/register/form/header/title.html.twig'
            subtitle:
                template: '@SyliusShop/account/register/form/header/subtitle.html.twig'
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/services/twig/component.xml [hunk 4]
```diff
@@ -141,5 +141,19 @@
                 template="@SyliusShop/product/show/page/info/summary/variant_selection/add_to_cart.html.twig"
             />
         </service>
+
+        <service
+            id="sylius_shop.twig.component.register.form"
+            class="Sylius\Bundle\ShopBundle\Twig\Component\Register\RegisterComponent"
```
Comment (GSadee) [inline src/Sylius/Bundle/ShopBundle/Resources/config/services/twig/component.xml]:
```suggestion
            class="Sylius\Bundle\ShopBundle\Twig\Component\Register\FormComponent"
```
or
```suggestion
            class="Sylius\Bundle\ShopBundle\Twig\Component\Account\RegisterFormComponent"
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/services/twig/component.xml [hunk 5]
```diff
@@ -141,5 +141,19 @@
...
+        >
+            <argument type="service" id="form.factory" />
+            <argument>Sylius\Bundle\CoreBundle\Form\Type\Customer\CustomerRegistrationType</argument>
+
+            <tag
+                name="sylius.live_component"
+                key="sylius_shop:register:form"
```
Comment (mpysiak) [inline src/Sylius/Bundle/ShopBundle/Resources/config/services/twig/component.xml]:
```suggestion
                key="sylius_shop:account:register:form"
```

File: src/Sylius/Bundle/ShopBundle/Resources/config/services/twig/component.xml [hunk 6]
```diff
@@ -155,5 +155,19 @@
                 template="@SyliusShop/product/show/page/product_listing/association.html.twig"
             />
         </service>
+
+        <service
+            id="sylius_shop.twig.component.register.form"
+            class="Sylius\Bundle\ShopBundle\Twig\Component\Register\FormComponent"
+        >
+            <argument type="service" id="form.factory" />
+            <argument>Sylius\Bundle\CoreBundle\Form\Type\Customer\CustomerRegistrationType</argument>
```
Comment (mpysiak) [inline src/Sylius/Bundle/ShopBundle/Resources/config/services/twig/component.xml]: Use formType from `ShopBundle`? 🤔

**Review by GSadee — APPROVED** [review]

**Review by mpysiak — APPROVED** [review]

File: src/Sylius/Behat/Context/Ui/Shop/LoginContext.php [hunk 7]
```diff
@@ -288,4 +299,12 @@ public function iShouldNotBeAbleToChangeMyPasswordWithThisToken(): void
     {
         Assert::false($this->resetPasswordPage->isOpen(), 'User should not be on the forgotten password page');
     }
+
+    /**
+     * @Then I should see who I am
+     */
+    public function iShouldSeeWhoIAm(): void
+    {
+        Assert::notSame('', $this->homePage->getFullName());
```
Comment (Rafikooo) [inline src/Sylius/Behat/Context/Ui/Shop/LoginContext.php]: It asserts only the existence of a name, but it does not guarantee that it's the correct one.

File: src/Sylius/Behat/Page/Shop/Account/LoginPage.php [hunk 8]
```diff
@@ -48,7 +48,7 @@ protected function getDefinedElements(): array
             'login_button' => '[data-test-button="login-button"]',
             'password' => '[data-test-login-password]',
             'username' => '[data-test-login-username]',
-            'validation_error' => '[data-test-flash-message="negative"]',
+            'validation_error' => '[data-test-sylius-flash-message]',
```
Comment (Rafikooo) [inline src/Sylius/Behat/Page/Shop/Account/LoginPage.php]:
```suggestion
            'flash_message' => '[data-test-sylius-flash-message]',
```
For the sake of consistency.

File: src/Sylius/Behat/Page/Shop/Product/IndexPage.php [hunk 9]
```diff
@@ -128,7 +128,7 @@ protected function getDefinedElements(): array
             'product' => '[data-test-product=%productCode%]',
             'products' => '[data-test-products]',
             'search_button' => '[data-test-search]',
-            'validation_message' => '[data-test-flash-message]',
+            'validation_message' => '[data-test-sylius-flash-message]',
```
Comment (Rafikooo) [inline src/Sylius/Behat/Page/Shop/Product/IndexPage.php]: As above.

File: src/Sylius/Bundle/CoreBundle/Form/EventSubscriber/CustomerRegistrationFormSubscriber.php [hunk 10]
```diff
@@ -33,29 +32,26 @@ public static function getSubscribedEvents(): array
...
-     * @throws \InvalidArgumentException
-     */
     public function preSubmit(FormEvent $event): void
     {
         $rawData = $event->getData();
         $form = $event->getForm();
         $data = $form->getData();
 
-        Assert::isInstanceOf($data, CustomerInterface::class);
+        if ($data instanceof CustomerInterface) {
```
Comment (Rafikooo) [inline src/Sylius/Bundle/CoreBundle/Form/EventSubscriber/CustomerRegistrationFormSubscriber.php]: Would be more readable if inverted.
```suggestion
        if (!$data instanceof CustomerInterface) {
            return;
        }
        ...
```

File: src/Sylius/Bundle/CoreBundle/Form/EventSubscriber/CustomerRegistrationFormSubscriber.php [hunk 11]
```diff
@@ -33,29 +32,26 @@ public static function getSubscribedEvents(): array
...
+            // if email is not filled in, go on
+            if (!isset($rawData['email']) || empty($rawData['email'])) {
+                return;
+            }
```
Comment (Rafikooo) [inline src/Sylius/Bundle/CoreBundle/Form/EventSubscriber/CustomerRegistrationFormSubscriber.php]:
```suggestion
            if (empty($rawData['email'])) {
                return;
            }
```
Should do the same work.

**Review by Rafikooo — APPROVED** [review]

**Pull request merged**